### PR TITLE
Add Ctrl+Backspace handler

### LINKED ISSUE #259: Create proper Ctrl+Backspace shortcut in TextBox
In Windows, `Ctrl+Backspace` commonly serves as a shortcut to delete the word immediately preceding the cursor. It is equivalent to `Ctrl+Shift+Left` followed by `Backspace`.

In WinForms, this shortcut works only in single-line `TextBox` controls which have autocomplete enabled. This is actually built into `SHAutoComplete` somewhere in the win32 API, and is a "rogue feature" with no documentation (some history [here](https://blogs.msdn.microsoft.com/oldnewthing/20071011-00/?p=24823)). If the `TextBox` is multi-line or doesn't have autocomplete on, `Ctrl+Backspace` inserts a DELETE character (ASCII 127) instead.

Note: The `Ctrl+Delete` shortcut (which deletes the word following the cursor) works in all relevant WinForms controls.

Outside of WinForms, the `Ctrl+Backspace` shortcut is nearly ubiquitous on Windows (as of October 2018, even Notepad finally has it!). As such, I propose properly implementing it in WinForms controls such as `TextBoxBase` and `ComboBox` with a custom handler for `Ctrl+Backspace`.

***

I've submitted a PR with one approach to add this shortcut, namely ~~using `SendInput`~~ sending native messages to simulate the aforementioned `Ctrl+Shift+Left, Backspace` keystrokes. I believe this is the right approach — as even the current autocomplete `TextBox` "feature" seems to be selecting the text and then deleting it if you watch closely.

Another possibility would be to use internal text selection functions rather than sending keystrokes. However, that requires replicating the logic dealing with word boundaries/punctuation currently used by `Ctrl+Shift+Left`, which is a bit quirky, not to mention wildly inconsistent between different Windows programs. We can consider this approach if the current PR is lacking.

### PR DESCRIPTION
**Pull request by ArrowCase**
Add new handler to `ProcessCmdKey` of `TextBoxBase` and `ComboBox` controls which intercepts `Ctrl+Backspace` (or `Ctrl+Shift+Backspace`) and sends keystrokes to select and delete the word immediately preceding the text cursor.

Previously, for [weird historical reasons](https://blogs.msdn.microsoft.com/oldnewthing/20071011-00/?p=24823), the win32 API implicitly gave `Ctrl+Backspace` this kind of behavior only in single-line, autocomplete-enabled `TextBox` controls. In other cases, it would insert an ASCII `DELETE` character instead.

If a text selection is present, only the selection is deleted. This behavior is consistent with the `Ctrl+Delete` shortcut.

(OUTDATED) Some notes on this implementation:
* Using `SendKeys` does not suffice.
    * If `^+{LEFT}{BACKSPACE}` were sent, the Control key would be considered released afterward. If the user continued holding Control and pressed Backspace a second time, it would act as a normal Backspace.
    * If `+{LEFT}{BACKSAPCE}` were sent (with the safe assumption of the Control key already being active), then nothing would actually happen until the user *released* the Control key.
* Since the Control key must be held to initiate this shortcut, we only send `Shift+Left`, not `Ctrl+Shift+Left`.
* Since the Control key is held, we can't send `Backspace` to do the deletion or we would trigger the shortcut recursively. `Delete` is sent instead, which actually becomes `Ctrl+Delete`. Luckily, the `Ctrl+Delete` shortcut already has the behavior of just deleting the current selection if one exists.
* `DeletePrecedingWord` is currently duplicated as I'm not sure of a good common location for it.

Resolves #259.

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs [hunk 1]
```diff
@@ -1476,6 +1476,60 @@ private string[] GetStringsForAutoComplete(IList collection)
...
+            inputs[3].inputUnion.ki.dwFlags = NativeMethods.KEYEVENTF_KEYUP;
+
+            inputs[4].type = NativeMethods.INPUT_KEYBOARD;
+            inputs[4].inputUnion.ki.wVk = (short)Keys.Delete;
+
+            inputs[5].type = NativeMethods.INPUT_KEYBOARD;
+            inputs[5].inputUnion.ki.wVk = (short)Keys.Delete;
+            inputs[5].inputUnion.ki.dwFlags = NativeMethods.KEYEVENTF_KEYUP;
+
+            UnsafeNativeMethods.SendInput((uint)inputs.Length, inputs, Marshal.SizeOf(typeof(NativeMethods.INPUT)));
```
Comment (sharwell) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs]: ❗️ This is scary to me for two reasons:

1. It makes the message handler for <kbd>Ctrl</kbd>+<kbd>Backspace</kbd> sensitive to the input focus, including race conditions where the operation could get sent to a totally different application if the user changes focus prior to the keyboard message getting handled.
2. It places the command at the back of the queue, allowing for reordering of Delete Previous Word relative to other typing commands if the user is typing faster than messages are getting processed.

If we want to fake the command via messaging, would it be possible to instead update `ProcessCmdKey` to directly call the base method multiple times using specially-crafted `Message` instances?

**Comment by ArrowCase** [comment]
@sharwell I was able to use `DispatchMessage()` instead of `SendInput()`. I can take the window handle from the existing `msg` parameter, so this ensures the messages are being sent to the right control/window.

However, it seems dispatching a `WM_KEYDOWN` message for the Shift key doesn't work; it's either ignored or immediately considered released.

So, the approach in the forthcoming comit is to call `UnsafeNativeMethods.SetKeyboardState()` to flag Shift as pressed before we send the `WM_KEYDOWN` for the left arrow, then call it again to unpress Shift afterward. If Shift is already pressed by the user, we skip those calls.

I also changed the existing text selection case, since clearing the selection with `this.SelectedText = "";` could not be undone by `Ctrl+Z`. This commit is `TextBoxBase` for now, but ideally this will be shared with `ComboBox`.

I can't speak the potential for race conditions with user input with this approach.

❓ One additional concern is the scan codes of the Left and Delete keys (part of the `lParam` of the message). They're hard coded right now, though Win32 provides a `MapVirtualKeyA` function which can map the virtual key code to a scan code. But would we in fact need to load the keyboard locale and pass it to `MapVirtualKeyExA` instead? Or can we rely on those keys having consistent scan codes across locales?

**Comment by zsd4yr** [comment]
What happens when you perform this operation on a native multiline text box? 

Trying to determine if this is something windows forms might want to handle on our end or if this is something Windows might want to look at.

**Comment by zsd4yr** [comment]
@ArrowCase can you make sure that there is no change in interaction under Narrator and Inspect (accessibility); any unexpected behaviors?

**Comment by ArrowCase** [comment]
I'll look into it.

**Comment by zsd4yr** [comment]
Hey @ArrowCase I wanted to check in and see how your regression testing is proceeding?

**Comment by wjk** [comment]
@ArrowCase I believe @zsd4yr is referring to a tool in the Windows 10 SDK called `Inspect.exe`. IIRC this tool displays in a list the accessibility properties of what's under the cursor, but it's been forever since I've last used it. You can find this tool in the `C:\Program Files (x86)\Windows Kits\10\bin\{version}\x86` folder, if it's not on your VS Developer Command Prompt path. Hope this helps!

**Comment by zsd4yr** [comment]
@ArrowCase please let us know what you find after using Inspect.exe @wjk indicated -- sorry about eh confusion there

**Comment by RussKie** [comment]
FWIW, in GitExtensions we had a similar issue and it was solved in a different manner
https://github.com/gitextensions/gitextensions/pull/6049

**Comment by ArrowCase** [comment]
I will post test results shortly.

@RussKie The problem with that approach I found (as an old workaround in my own WinForms applications) is that if you hold Ctrl and press Backspace multiple times, only the first one results in the shortcut working properly, and subsequent presses are treated as a plain Backspace.

**Comment by RussKie** [comment]
> if you hold Ctrl and press Backspace multiple times, only the first one results in the shortcut working properly, and subsequent presses are treated as a plain Backspace.

Thank you! I could repro the issue just like you described it. I wonder how we missed it while testing...

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs [hunk 2]
```diff
@@ -1476,6 +1476,60 @@ private string[] GetStringsForAutoComplete(IList collection)
...
+
+            return returnValue;
+        }
+
+        /// <devdoc>
+        ///     Send Ctrl+Shift+Left, Delete to delete the word preceding the cursor.
+        ///     This is called when Ctrl+Backspace is pressed, so the Ctrl key is always active.
+        /// </devdoc>
+        /// <internalonly/>
+        private void DeletePrecedingWord() {
```
Comment (wjk) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs]: Is there any reason why we have to use `SendInput` here, while we can use `WM_KEYDOWN` in `TextBoxBase`? AFAICT the ComboBox control should also support the same selection sequence natively.

File: src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs (hunk 1 above)
Comment (wjk) [inline src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.cs]: I'd recommend using `Marshal.SizeOf<NativeMethods.INPUT>()` here, as the non-generic overload was deprecated starting in .NET Framework 4.5.1.

**Comment by zsd4yr** [comment]
@sharwell are you satisfied with the changes that have been made here in response to your feedback? 

@ArrowCase have you had an opportunity to address @wjk 's concerns?

**Comment by sharwell** [comment]
@zsd4yr No, `SendInput` is inherently a problem. Note also that state tracking for `SendInput` can be problematic, leading to inconsistencies between the OS understanding of which modifiers are pressed vs. the application's understanding. This caused a bunch of problems for Roslyn integration testing and we had to work around it by using scan codes instead of virtual keys (https://github.com/dotnet/roslyn/pull/33524).

I can remove my block if you want, but I couldn't approve a change that resulted in the use of the `SendInput` API being used by a production code feature. There are just too many ways it can go wrong.

**Comment by ArrowCase** [comment]
@sharwell The intent is to avoid SendInput entirely. The PR is just in a half-baked state with the changes to ComboBox not up to date with those to TextBoxBase. Can you please comment on the approach taken for TextBoxBase, using DispatchMessage instead of SendInput?

**Comment by sharwell** [comment]
@ArrowCase DispatchMessage addresses the main concerns I had. I don't know the details of the API well enough to say if there are other problems with it. The main edge cases it would make sense to test are:

1. Message filters ([`Application.AddMesageFilter`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.application.addmessagefilter))
2. Overridden message handling in user code

**Comment by weltkante** [comment]
not a fan of faked WM_KEYDOWN messages which don't respect the [documention](https://docs.microsoft.com/en-us/windows/desktop/inputdev/wm-keydown) of the message, in particular on first glance it doesn't set the scancode and doesn't set the previous state bit. As long as they only get processed by the OS it wouldn't be a problem, but we have for example message filters in our app to implement an onscreen keyboard. How would this behave, would it see the DispatchMessage calls? How would it interact with the keyboard state of the UI thread?

On another note, did anyone test how these techniques interact with IME editors?

**Comment by weltkante** [comment]
Also its just a WM_KEYDOWN and no corresponding WM_KEYUP or am I reading that wrong? That definitely would confuse our implementation of the onscreen keyboard, leaving the key stuck in "pressed" visual state. (assuming the message filter sees those messages)

**Comment by ArrowCase** [comment]
@weltkante Thanks for the insight about the problems messages could cause. I wasn't thinking of message handlers beyond the OS, which could definitely cause problems.

All of this has convinced me that using messaging/input sending APIs is not a workable approach here. I'll close this PR for now and will submit another one that instead takes the approach of using text selections only, with no messaging. This will require implementing the logic that determines what a "word" is for the purpose of selection, but I don't think that's ultimately very complicated.
